### PR TITLE
Matching

### PR DESCRIPTION
**Pull request by ljwolf**
Hello! Please make sure to check all these boxes before submitting a Pull Request
(PR). Once you have checked the boxes, feel free to remove all text except the
justification in point 5. 

1. [x] You have run tests on this submission locally using `pytest` on your changes. Continuous integration will be run on all PRs with [GitHub Actions](https://github.com/pysal/libpysal/blob/master/.github/workflows/unittests.yml), but it is good practice to test changes locally prior to a making a PR.
2. [x] This pull request is directed to the `pysal/main` branch.
3. [x] This pull introduces new functionality covered by
   [docstrings](https://en.wikipedia.org/wiki/Docstring#Python) and
   [unittests](https://docs.python.org/2/library/unittest.html)? 
4. [x] You have [assigned a
   reviewer](https://help.github.com/articles/assigning-issues-and-pull-requests-to-other-github-users/) and added relevant [labels](https://help.github.com/articles/applying-labels-to-issues-and-pull-requests/)
5. [x] The justification for this PR is contained in #665 

# todo:

- [x] add `pulp` to our testing matrix. 
- [x] write tests using the toy example in [the gist](https://gist.github.com/ljwolf/1529b5cf5bfac2341fd8f103b53f9ad6)
- [x] write an example notebook

### CODE REVIEW

File: libpysal/graph/_matching.py [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+    return_mip : bool (default: False)
+        whether or not to return the instance of the pulp.LpProblem. By
+        default, the problem is not returned to the user.
+    allow_partial_match : bool (default: False)
+        whether to allow for partial matching. A partial match may have
+        a weight between zero and one, while a "full" match (by default)
+        must have a weight of either zero or one. A partial matching may
+        have a shorter total distance, but will result in a weighted
+        graph.
+    p : int (default: 2)
```
Comment (ljwolf) [inline libpysal/graph/_matching.py:62]: This isn't being sent to metric at the moment.

File: libpysal/graph/_matching.py [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+                    match_vars[j, i]
+                    for i in range(n_sources)
+                ]
+            )
+            sense = 1
+        else:
+            summand = pulp.lpSum(
+                [
+                    # in the "upper" triangle, or "lower" triangle
+                    match_vars.get((i, j), match_vars.get((j, i)))
```
Comment (ljwolf) [inline libpysal/graph/_matching.py:133]: The problem is much more efficient when `x=y`, and we can implement this trick to halve the number of decision variables. If we recursively solved `_spatial_matching(x=x, y=x)`, we'd lose this optimisation. I *think* this optimisation also would work if `x.shape==y.shape`.

Comment (ljwolf) [reply libpysal/graph/_matching.py:133]: No, the only reason why the upper triangle works is because the rows are exchangeable with the columns. If you had a 3 x 5 matching matrix, neither triangle can express the full set of possible matches.

All cases where x != y need the full assignment matrix, and we can only use the upper/lower triangle if y=x.

**Comment by ljwolf** [comment]
Tests are now added, but pulp installation fails on windows through `conda-forge`. Is there a better way to do this?

**Comment by jGaboardi** [comment]
> Tests are now added, but pulp installation fails on windows through `conda-forge`. Is there a better way to do this?

In `spopt` we install `pulp` via `pip` --> https://github.com/pysal/spopt/blob/4760bd9e03f9b31c003bfa7f5d785a2a81b174bd/ci/312-latest.yaml#L32

**Comment by jGaboardi** [comment]
Looks like `numpy` is never being imported --> https://github.com/pysal/libpysal/actions/runs/7159612453/job/19493044782?pr=666#step:7:2098

**Comment by jGaboardi** [comment]
Is [this](https://github.com/pysal/libpysal/actions/runs/7159753204/job/19493357426?pr=666#step:7:1796) passing locally for you?

**Comment by ljwolf** [comment]
Yes, but fixing that, I still have fails locally. Will fix.

**Comment by ljwolf** [comment]
One question I had for maintainers: 

I built the spatial matching framework to be able to do cross-pattern queries. So, something like: 

> For all points in pattern X, find at least k matched points in pattern Y such that the total distance of X->Y matches are minimised. 

I think this would be good to expose in `cg`, since it's a general purpose spatial matching function for computational geometry applications. I have something even more general that will probably make it into a separate package for spatial causal inference in python, but the spatial implementation can live here. 

So, would it be OK to grab `_spatial_matching()` from `libpysal.graph._matching` and expose it directly in `libpysal.cg`? I could alternatively define a dispatching function that *requires* the second pattern (`y` argument) to be provided.

File: libpysal/graph/_matching.py [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+                ]
+            )
+            sense = int(not allow_partial_match)
+        mp += pulp.LpConstraint(summand, sense=sense, rhs=n_matches)
+    if match_between:  # but, we may choose to ignore some sources
+        for i in range(n_sources):
+            summand = pulp.lpSum([match_vars[j, i] for j in range(n_targets)])
+            mp += pulp.LpConstraint(summand, sense=-1, rhs=n_matches)
+    if solver is None:
+        solver = pulp.COIN(msg=False)
```
Comment (jGaboardi) [inline libpysal/graph/_matching.py:151]: I'm wondering if there is way to simply choose the default installed solver without needing to specific (or is this what's happening here with `pulp.COIN()`?). Implementing that logic would go a long way over in `spopt` >> https://github.com/pysal/spopt/issues/348

cc @gegen07

Comment (ljwolf) [reply libpysal/graph/_matching.py:149]: I think we need to do this, since `pulp.PULP_CBC_CMD()` I think is different from `pulp.COIN_CMD()`, and some students could run `pulp.GUROBI()` but not `pulp.GUROBI_CMD()`. 

I think we should really leave it to pulp to deal with this, rather than specifying COIN.

Comment (jGaboardi) [reply libpysal/graph/_matching.py]: Yeah, seems like it would make a nice improvement in `spopt` to have the ability to pass in no solver, and then the default one installed is used [within the `solve()` methods](https://github.com/pysal/spopt/blob/ec36c133fbef6f1cde1b3974d34e69b703eed03a/spopt/locate/p_dispersion.py#L315).

**Comment by ljwolf** [comment]
The code is ready to review. I'm not sure why the ruff check is failing... I've run the ruff auto formatter myself locally on the submission (2c04acf, 215022d, 273e6ea)

**Comment by jGaboardi** [comment]
> The code is ready to review. I'm not sure why the ruff check is failing... I've run the ruff auto formatter myself locally on the submission ([2c04acf](https://github.com/pysal/libpysal/commit/2c04acf749ef12c10536e164af95437fbc7e61d4), [215022d](https://github.com/pysal/libpysal/commit/215022db802949c4f2dcadca53d858f2bd594529), [273e6ea](https://github.com/pysal/libpysal/commit/273e6ea27000b8540da604272fb75e34de4cdc7f))

Do you have the most recent `ruff` installed locally? I think there was a fresh version released last night. Could be playing a part.

**Comment by ljwolf** [comment]
Yes, not quite sure what's up here: 

Screenshot 2023-12-18 at 11 42 40

0.1.8 is the most recent version of ruff, and it fixes nothing both in the CI and locally?

**Comment by martinfleis** [comment]
It is `ruff-format` that is failing, not `ruff` linter.

**Comment by ljwolf** [comment]
OK! Wasn't clear on the differences between ruff check and format. Sorry! 

I think this is complete.

File: libpysal/graph/_matching.py [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+        for i in range(n_sources):
+            summand = pulp.lpSum([match_vars[j, i] for j in range(n_targets)])
+            mp += pulp.LpConstraint(summand, sense=-1, rhs=n_matches)
+
+    status = mp.solve(solver)
+
+    if (status != 1) & (not allow_partial_match):
+        warnings.warn(
+            f"Problem is {pulp.LpStatus[status]}, so edge weights may be non-integer!",
+            stacklevel=1,
```
Comment (jGaboardi) [inline libpysal/graph/_matching.py:155]: This should probably be `stacklevel=2`, if I'm not mistaken.

File: libpysal/graph/tests/test_matching.py [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+    index_parts=False
+)
+stores_unique = stores.drop_duplicates(subset="geometry")
+
+np.random.seed(85711)
+simple = np.random.random(size=(5, 2))
+
+
+def test_correctness_k1():
+    pulp = pytest.importorskip("pulp")
```
Comment (jGaboardi) [inline libpysal/graph/tests/test_matching.py:29]: Let's move this above and skip all tests in the file if `pulp` isn't installed.

File: libpysal/graph/tests/test_matching.py [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+    pytest.importorskip("sklearn")
+    default_solver = pulp.listSolvers(onlyAvailable=True)
+    if len(default_solver) == 0:
```
Comment (jGaboardi) [inline libpysal/graph/tests/test_matching.py:32]: Same for fetching and defining the default solver. This only needs to happen once.

File: libpysal/graph/tests/test_matching.py (hunk 5 above)
Comment (jGaboardi) [reply libpysal/graph/tests/test_matching.py:29]: And same with `sklearn`

**Review by jGaboardi — CHANGES_REQUESTED** [review]
Looking very cool.

File: ci/310-oldest.yaml [hunk 7]
```diff
@@ -29,6 +29,7 @@ dependencies:
   - sqlalchemy=2.0
   - zstd
   - xarray=2022.3
-  - pip
+  - pulp
```
Comment (martinfleis) [inline ci/310-oldest.yaml:32]:
```suggestion
  - pip
```

File: ci/312-no-optional.yaml [hunk 8]
```diff
@@ -14,6 +14,7 @@ dependencies:
   # testing
   - codecov
   - matplotlib
+  - scikit-learn
```
Comment (martinfleis) [inline ci/312-no-optional.yaml:17]:
```suggestion
```
No optional dependencies shall be in this env

File: environment.yml [hunk 9]
```diff
@@ -26,3 +26,7 @@ dependencies:
   - sqlalchemy
   - xarray
   - zstd
+  - coin-or-cbc
```
Comment (martinfleis) [inline environment.yml:29]: Why do we have this here and not in CI envs?

File: libpysal/graph/_matching.py [hunk 10]
```diff
@@ -0,0 +1,168 @@
+import warnings
+
+import numpy
+from sklearn.metrics import pairwise_distances
```
Comment (martinfleis) [inline libpysal/graph/_matching.py:4]: So far, sklearn has been a soft dependency for libpysal. Do we want to change that?

**Review by martinfleis — COMMENTED** [review]
The large diff in base is just moving stuff around to get alphabetic order of builders?

Can you also add a test for `Graph.build_spatial_matches` ?

File: ci/312-no-optional.yaml (hunk 8 above)
Comment (jGaboardi) [reply ci/312-no-optional.yaml:17]: Good catch @martinfleis

**Comment by jGaboardi** [comment]
> The large diff in base is just moving stuff around to get alphabetic order of builders?
> 
> Can you also add a test for `Graph.build_spatial_matches` ?

Yes.

File: libpysal/graph/_matching.py (hunk 10 above)
Comment (knaaptime) [reply libpysal/graph/_matching.py:4]: yes, we decided to change

File: ci/312-no-optional.yaml (hunk 8 above)
Comment (knaaptime) [reply ci/312-no-optional.yaml:17]: made it a hard dep

Comment (ljwolf) [reply ci/312-no-optional.yaml:17]: So, I will keep `scikit-learn` in this env?

File: environment.yml (hunk 9 above)
Comment (ljwolf) [reply environment.yml:29]: This was part of my attempt to get pulp installed correctly. Fine to remove.

File: ci/312-no-optional.yaml (hunk 8 above)
Comment (martinfleis) [reply ci/312-no-optional.yaml:17]: I suppose so but it needs to make its way to requirements in pyproject.toml if it is becoming hard dep now.

**Comment by ljwolf** [comment]
I believe I've addressed all comments!

**Review by knaaptime — APPROVED** [review]

**Review by jGaboardi — APPROVED** [review]

**Comment by ljwolf** [comment]
Hi! could one of the approvers please merge this?

**Comment by ljwolf** [comment]
thank you!
